*** xpath.c	2001/02/18 16:49:41	1.74
--- xpath.c	2001/02/19 09:12:00
*************** xmlXPathEvalPathExpr(xmlXPathParserConte
*** 5149,5154 ****
--- 5149,5157 ----
      if ((CUR == '$') || (CUR == '(') || (IS_DIGIT(CUR)) ||
          (CUR == '\'') || (CUR == '"')) {
  	lc = 0;
+     } else if (CUR == '*') {
+ 	/* relative or absolute location path */
+ 	lc = 1;
      } else if (CUR == '/') {
  	/* relative or absolute location path */
  	lc = 1;
